### PR TITLE
refactor namespace to kind of target

### PR DESCRIPTION
**Pull request by neo-liang-sap**
**What this PR does / why we need it**:
This PR makes namespace to be one kind of target, just like garden/seed/project/shoot

**Which issue(s) this PR fixes**:
Fixes # https://github.com/gardener/gardenctl/issues/169

**Special notes for your reviewer**:
* `gardenctl ls namespace` is supported
* i've tested in following scenario/combination
```
gardenctl target -g canary-virtual -s gcp-eu2 -p everest -t ev-1373-b -n kube-system
gardenctl target -g canary-virtual -s gcp-eu2 -p everest -n kube-system
gardenctl target -g canary-virtual -s gcp-eu2 -n kube-system
gardenctl target -g canary-virtual -n kube-system
gardenctl target -s gcp-eu2 -p everest -t ev-1373-b -n kube-system
gardenctl target -s gcp-eu2 -t ev-1373-b -n kube-system
gardenctl target -s gcp-eu2 -n kube-system
gardenctl target -p everest -t ev-1373-b -n kube-system
gardenctl target -p everest -n kube-system
gardenctl target -n kube-system

gardenctl target -g canary-virtual -p everest -t ev-1373-b -n kube-system
gardenctl target -g canary-virtual -p everest -n kube-system
gardenctl target -g canary-virtual -t ev-1373-b -n kube-system

gardenctl target -t ev-1373-b -n kube-system
gardenctl target -p everest -n kube-system
gardenctl target -s gcp-eu2 -n kube-system

gardenctl target namespace kube-system
```

and verified target status in 
`cat /Users/i352986/.garden/sessions/plantingSession/target`
**Release note**:

```improvement operator
Namespace is now supported as one kind of target
gardenctl drop namespace is now supported
gardenctl ls namespaces is now supported
```

### CODE REVIEW

**Comment by DockToFuture** [comment]
`Gardenctl drop` is currently removing the ns from the stack but not from the kube-config.
And when I'm targeting another shoot i get the kubeconfig printed of the new shoot but I'm still targeting the same shoot.
E.g. target stack is:
```
target:
- kind: garden
  name: dev
- kind: project
  name: core
- kind: shoot
  name: cli-gcp
- kind: namespace
  name: kube-system
```
after targeting a new shoot I still have the same shoot in the target stack
```
$ g target shoot ts-test
Shoot:
KUBECONFIG=/Users/xxx/.garden/cache/dev/projects/core/ts-test/kubeconfig.yaml
xxx at vvv in ~/go/src/github.com/gardener/gardenctl on pr-173
$ g get target
target:
- kind: garden
  name: dev
- kind: project
  name: core
- kind: shoot
  name: cli-gcp
- kind: namespace
  name: kube-system
```

**Comment by neo-liang-sap** [comment]
Hi @DockToFuture ,

I fixed second issue in https://github.com/gardener/gardenctl/pull/173/commits/c86ab58a6c39ae39688dbeb38e709a6468e9ca11 , `targetShoot` miss logic when current target.Target size is 4 (including namespace information)

For the first issue `Gardenctl drop is currently removing the ns from the stack but not from the kube-config.` sorry i didn't quite get the point, currently as namespace doesn't have `kubeconfig` information so i put the logic in https://github.com/gardener/gardenctl/pull/173/commits/b6a0bb00213b1e5728accda883ee1115ec4febce#diff-d0d3bde7a059e586e30272e957484afdR633 to https://github.com/gardener/gardenctl/pull/173/commits/b6a0bb00213b1e5728accda883ee1115ec4febce#diff-d0d3bde7a059e586e30272e957484afdR648 and this function `getKubeConfigOfCurrentTarget` is called in `drop()` function.

Could you please raise an example for this issue? then i can get an better understandinfg on this...

Thanks a lot!
-Neo

**Comment by DockToFuture** [comment]
I would expect in the following example, that after I have dropped the namespace from the target stack and I execute `g k get pods` that I get the pods in the default ns listed. Therefore `gardenctl drop` has to unset the namespace. WDYT?

```
$ g get target
target:
- kind: garden
  name: dev
- kind: project
  name: core
- kind: shoot
  name: cli-gcp
- kind: namespace
  name: kube-system
xxx at yyy in ~/go/src/github.com/gardener/gardenctl on pr-173
```
```
$ g drop
Dropped namespace kube-system
xxx at yyy in ~/go/src/github.com/gardener/gardenctl on pr-173
```

```
$ g get target
target:
- kind: garden
  name: dev
- kind: project
  name: core
- kind: shoot
  name: cli-gcp
xxx at yyy in ~/go/src/github.com/gardener/gardenctl on pr-173
```
```
$ g k get pods
NAME                                                             READY   STATUS    RESTARTS   AGE
addons-nginx-ingress-controller-5796c65bf8-cchc7                 1/1     Running   0          2d23h
addons-nginx-ingress-nginx-ingress-k8s-backend-b7c567f7b-nq2lg   1/1     Running   0          2d23h
blackbox-exporter-58cf74dd57-xlq75                               1/1     Running   0          2d23h
calico-kube-controllers-6f864659cf-plw7f                         1/1     Running   0          2d23h
calico-node-hj4wq                                                1/1     Running   0          8h
calico-node-nhj8f                                                1/1     Running   0          8h
calico-typha-deploy-6c5dc7856f-vs2s9                             1/1     Running   0          2d23h
calico-typha-horizontal-autoscaler-6455dc88c9-7fljk              1/1     Running   0          2d23h
calico-typha-vertical-autoscaler-7b9954b9df-962xs                1/1     Running   0          2d23h
coredns-db5c59c87-76pnc                                          1/1     Running   0          2d23h
coredns-db5c59c87-v92jw                                          1/1     Running   0          2d23h
kube-proxy-hhnxm                                                 1/1     Running   0          8h
kube-proxy-zptfc                                                 1/1     Running   0          8h
metrics-server-5d5788576f-t96fw                                  1/1     Running   0          2d23h
node-exporter-8wfqs                                              1/1     Running   0          8h
node-exporter-qwhqh                                              1/1     Running   0          8h
node-problem-detector-kpt7m                                      1/1     Running   0          8h
node-problem-detector-t2p2f                                      1/1     Running   0          8h
vpn-shoot-7fbc9c65b8-t8qt7                                       1/1     Running   0          2d23h
```

**Comment by neo-liang-sap** [comment]
Hi @DockToFuture ,
Thanks for the comments!
I added some logic in `drop.go` - setting namespace to `default` while dropping an namespace
Thanks
-Neo

File: pkg/cmd/target.go [hunk 1]
```diff
@@ -628,8 +641,24 @@ func getKubeConfigOfClusterType(clusterType TargetKind) (pathToKubeconfig string
 
 // getKubeConfigOfCurrentTarget returns the path to the kubeconfig of current target
 func getKubeConfigOfCurrentTarget() (pathToKubeconfig string) {
+	var targetReal Target
 	var target Target
-	ReadTarget(pathTarget, &target)
+	ReadTarget(pathTarget, &targetReal)
+
+	if len(targetReal.Target) == 1 && targetReal.Stack()[0].Kind == "namespace" {
+		panic("the target has only namespace, this is invalid, at least one garden needs to be targeted before using namespace")
```
Comment (DockToFuture) [inline pkg/cmd/target.go]: Can you replace `panic` function with `return errors.New("error message")`? User will be confused by the stack trace.

File: pkg/cmd/target.go [hunk 2]
```diff
@@ -628,8 +641,24 @@ func getKubeConfigOfClusterType(clusterType TargetKind) (pathToKubeconfig string
...
+	} else if len(targetReal.Target) > 1 && len(targetReal.Target) <= 4 {
+		if targetReal.Stack()[len(targetReal.Target)-1].Kind == "namespace" {
+			target.Target = targetReal.Target[:len(targetReal.Target)-1]
+		} else {
+			target.Target = targetReal.Target
+		}
+	} else if len(targetReal.Target) == 1 && targetReal.Stack()[0].Kind != "namespace" {
+		target.Target = targetReal.Target
+	} else {
+		panic("length of target.Stack is illegal")
```
Comment (DockToFuture) [inline pkg/cmd/target.go]: see above

File: pkg/cmd/target.go [hunk 3]
```diff
@@ -789,3 +841,48 @@ func setKubectlNamespace(kubectlNameSpace string) error {
...
 	return nil
 }
+
+//write current namespace to target
+func targetNamespace(targetWriter TargetWriter, ns string) error {
+	var target Target
+	ReadTarget(pathTarget, &target)
+
+	if len(target.Target) > 4 {
+		panic("the length is greater than 4 and illegal")
```
Comment (DockToFuture) [inline pkg/cmd/target.go]: see above

File: pkg/cmd/target.go [hunk 4]
```diff
@@ -789,3 +841,48 @@ func setKubectlNamespace(kubectlNameSpace string) error {
...
+	}
+	if len(target.Target) == 0 {
+		panic("the length is 0 and illegal. at least one garden needs to be targeted")
```
Comment (DockToFuture) [inline pkg/cmd/target.go]: see above

File: pkg/cmd/target.go [hunk 5]
```diff
@@ -789,3 +841,48 @@ func setKubectlNamespace(kubectlNameSpace string) error {
...
+	}
+	if len(target.Target) == 1 {
+		if string(target.Target[0].Kind) != "garden" {
+			panic("if one element in target, this needs to be garden")
```
Comment (DockToFuture) [inline pkg/cmd/target.go]: see above

**Comment by neo-liang-sap** [comment]
Hi @DockToFuture ,
thanks, for three occurence of panic i used return errors.New("") to replace.
For two occurence which function doesn't return error but only kubeconfigpath , i use fmt.Println("") to print error message and os.Exit(2) to fail the process

Thanks!

**Comment by neo-liang-sap** [comment]
thanks @DockToFuture !

**Pull request merged**